lib/Dialect/TritonGPU/Transforms/WarpSpecialization/OptimizePartitionWarps.cpp

```diff
@@ -249,16 +249,23 @@ static LogicalResult optimizePartitionNumWarps(ModuleAxisInfoAnalysis &axisInfo,
     }
   } while (changed);
 
-  for (auto [partition, newNumWarps, prevNumWarps, tensorRegs] :
+  SmallVector<int32_t> estRegUsage(partitionNumWarps.size());
+  for (auto [partition, newNumWarps, prevNumWarps, tensorRegs, estRegs] :
        llvm::zip(wsOp.getPartitionRegions(), partitionNumWarps,
-                 wsOp.getPartitionNumWarps(), maxTensorRegs)) {
+                 wsOp.getPartitionNumWarps(), maxTensorRegs, estRegUsage)) {
+    // "Guess" the register usage for each partition.
+    estRegs = tensorRegs ? 80 : 48;
```

Comment:
Where these magic numbers come from?

Reply:
Profiles... It's practically impossible to estimate register usage from TTGIR without compiling the code with PTXAS.

Reply:
But they are tight to specific ML code, right? If I change the triton kernel, the number of registers won't be 80 or 48 any more

Reply:
In this case, these are the register usages of code that warp specialize specifically puts inside partitions, so the compiler has more control over what that looks like. User code remains in the default partition and it gets whatever extra registers are freed up.

Reply:
I see. So we are under the assumption that compiler will likely choose either 80 or 48?

Reply:
Yes. 48 for arithmetic only partitions. In practice the arithmetic partitions use a max of 25-28 registers, but if I choose 32 for some reason it starts to spill. No idea what PTXAS is doing here, but 48 works well enough.